File: cpp/ampls/examples/benders.cpp
#include "ampls/ampls.h"
#include "test-config.h" // for MODELS_DIR
#include "ampl/ampl.h"

#include <vector>
#include <map>
#include <limits>
#include <algorithm>>
#include <iostream>
#include <iomanip>
#include <cassert>

// Utility function to extract keys and values from a map<string, double>
std::pair<std::vector<ampl::Tuple>, std::vector<double>> to_tuple(const std::map<std::string, double>& myMap) {
  std::vector<ampl::Tuple> keys;
  std::vector<double> values;

  for (const auto& pair : myMap) {
    keys.push_back(ampl::Tuple(pair.first));
    values.push_back(pair.second);
  }

  return std::make_pair(keys, values);
}

// Utility function to extract keys and values from a map<string, map<string, double>>
std::pair<std::vector<ampl::Tuple>, std::vector<double>> to_tuple(const
  std::map<std::string, std::map<std::string, double>>& myMap) {
  std::vector<ampl::Tuple> keys;
  std::vector<double> values;

  for (const auto& outerPair : myMap) {
    for (const auto& innerPair : outerPair.second) {
      keys.push_back(ampl::Tuple(innerPair.first, outerPair.first));
      values.push_back(innerPair.second);
    }
  }

  return std::make_pair(keys, values);
}

void create_common(ampl::AMPL& a) {
  a.eval("set FACILITIES;"
    "set CUSTOMERS;"
    "set SCENARIOS;"
    "param prob{SCENARIOS} default 1/card(SCENARIOS);"
    "param sub_scenario symbolic in SCENARIOS;"
    "param customer_demand{CUSTOMERS, SCENARIOS} >= 0;"

    "param facility_capacity{FACILITIES} >= 0;"
    "param variable_cost{ FACILITIES, CUSTOMERS } >= 0;"

  );
}

void create_sub_problem(ampl::AMPL& a) {
  create_common(a);
  a.eval(
    "param sub_facility_open{ FACILITIES } default 1; "
    "var production{FACILITIES, CUSTOMERS, SCENARIOS} >= 0;"
    "minimize operating_cost:"
    "sum{ i in FACILITIES, j in CUSTOMERS }"
    "variable_cost[i, j] * production[i, j, sub_scenario]; "

    "s.t.satisfying_customer_demand{ j in CUSTOMERS }:"
    "sum{i in FACILITIES} production[i, j, sub_scenario] >= customer_demand[j, sub_scenario];"

    "s.t.facility_capacity_limits{i in FACILITIES}:"
    "sum{ j in CUSTOMERS } production[i, j, sub_scenario] <= facility_capacity[i] * sub_facility_open[i];");
}

void create_master_problem(ampl::AMPL& a) {
  create_common(a);
  a.eval("param fixed_cost{ FACILITIES } >= 0;"
    "var sub_variable_cost{SCENARIOS} >= 0;"
    "var facility_open{ FACILITIES } binary;"
    "minimize TotalCost :"
    "sum{ i in FACILITIES } fixed_cost[i] * facility_open[i] + sum{s in SCENARIOS} prob[s]*sub_variable_cost[s];"
    "s.t. sufficient_production_capacity:"
    "sum{ i in FACILITIES } facility_capacity[i] * facility_open[i] >= max{ s in SCENARIOS } sum{ j in CUSTOMERS } customer_demand[j, s];");
}

// Load data into the AMPL model
void load_data(ampl::AMPL& a, const std::vector<const char*>& facilities,
  const std::vector<const char*>& customers, const std::vector<const char*>& scenarios,
  const std::map<std::string, double>& facility_capacity_map,
  const std::map<std::string, std::map<std::string, double>>& customer_demand_map,
  const std::map<std::string, std::map<std::string, double>>& variable_cost_map
) {

  a.getSet("FACILITIES").setValues(facilities.data(), facilities.size());
  a.getSet("CUSTOMERS").setValues(customers.data(), customers.size());
  a.getSet("SCENARIOS").setValues(scenarios.data(), scenarios.size());

  auto fc = to_tuple(facility_capacity_map);
  a.getParameter("facility_capacity").setValues(std::get<0>(fc).data(), std::get<1>(fc).data(), std::get<0>(fc).size());
  auto vc = to_tuple(variable_cost_map);
  a.getParameter("variable_cost").setValues(std::get<0>(vc).data(), std::get<1>(vc).data(), std::get<0>(vc).size());

  auto cd = to_tuple(customer_demand_map);
  a.getParameter("customer_demand").setValues(std::get<0>(cd).data(), std::get<1>(cd).data(), std::get<0>(cd).size());
}
void load_master_data(ampl::AMPL& a) {
  a.eval("data;"
    "param fixed_cost :=   Watson 39268000 Bayshore 65268000 Orange 60268000 Evanston 48268000;");
}


int findIndexWithSubstring(const std::map<std::string, int>& myMap, const std::string& substring) {
  for (const auto& pair : myMap) {
if (pair.first.find(substring) != std::string::npos) {
  return pair.second; // Return the integer associated with the key
}
    }
  throw std::runtime_error("Index not found!");
}

/** Add a cut (optimality of feasibility) to the model*/
void addBendersCut(ampls::AMPLModel& model, const ampl::DataFrame& cdduals,
  const ampl::DataFrame& fcduals, 
  const std::map<std::string, int>& index_facility_open,
  int index_sub_variable_cost_scen,
  const std::map<std::string, double>& facility_capacity,
  const std::map<std::string, double>& scenario_customer_demand,
  bool optimalityCut) {

  // Note that to accomodate for solver view, we formulate it as below
  // sub_variable_cost[s] - sum{ i in FACILITIES } facility_price[i, s, k] * facility_capacity[i] * facility_open[i]
  // >=   sum{j in CUSTOMERS} customer_price[j,s,k]*customer_demand[j,s]; 
  std::vector<int> indices;
  std::vector<double> coeffs;
  for (auto fc : fcduals) {
      indices.push_back(index_facility_open.at(fc[0].str()));
      coeffs.push_back(-fc[1].dbl() * facility_capacity.at(fc[0].str()));
  }
  if (optimalityCut) { 
    indices.push_back(index_sub_variable_cost_scen);
    coeffs.push_back(1);
  }
  double rhs = 0;
  for (auto cd : cdduals) {
    rhs += cd[1].dbl() * scenario_customer_demand.at(cd[0].str());
  }
  std::string name = optimalityCut ? "optimality" : "feasibility";
  auto cut = model.addConstraint(indices.size(), indices.data(), coeffs.data(),
    ampls::CutDirection::GE, rhs);
  model.record(cut);
  std::cout << "Added " << name << " cut:\n" << cut.toString();
}

template <class T> void solve(ampl::AMPL& master, ampl::AMPL& sub) {

  // Set some options
  sub.eval("suffix dunbdd;");
  master.setIntOption("presolve", 0);
  sub.setIntOption("presolve", 0);
  sub.setIntOption("solver_msg", 0);

  // Data
  std::vector<const char*> FACILITIES({ "Watson", "Bayshore", "Orange", "Evanston"});
  std::vector<const char*> CUSTOMERS({ "CEM_Store_41594", "CEM_Cypress_Market", "CEM_Time_Mart_23", "CEM_Cypress_CEM", "CEM_Fuel_Maxx_84", "CEM_Commercial_CEM", "CEM_New_Brunswick", "CEM_Bernards_PGroup" });
  std::vector < const char*> SCENARIOS({ "Low", "Medium", "High" });
  std::map<std::string, double> facility_capacity_map{
        {"Watson", 75000},
        {"Bayshore",200000},
        {"Orange", 150000},
        {"Evanston", 90000}
  };

  std::map<std::string, double> lowDemand, mediumDemand, highDemand;
  for (auto c : CUSTOMERS) {
    lowDemand[c] = 10000;
    mediumDemand[c] = 15000;
    highDemand[c] = 20000;
  }

  std::map<std::string, std::map<std::string, double>> customer_demand_map{
      { "Low",  lowDemand},
      { "Medium",  mediumDemand },
      { "High",  highDemand }
  };

  std::map<std::string, std::map<std::string, double>> variable_cost_map{
        {"CEM_Store_41594", { {"Watson", 6739.72500}, {"Bayshore", 10355.05000}, {"Orange", 7650.40000}, {"Evanston", 5219.50000} }},
        {"CEM_Cypress_Market", { {"Watson", 4739.72500}, {"Bayshore", 8355.05000}, {"Orange", 17320.40000}, {"Evanston", 15000.50000} }},
        {"CEM_Time_Mart_23", { {"Watson", 7739.72500}, {"Bayshore", 9355.05000}, {"Orange", 7320.40000}, {"Evanston", 5000.50000} }},
        {"CEM_Cypress_CEM", { {"Watson", 5739.72500}, {"Bayshore", 6355.05000}, {"Orange", 7320.40000}, {"Evanston", 9433.70000}}},
        {"CEM_Fuel_Maxx_84", { {"Watson", 9739.72500}, {"Bayshore", 10355.05000}, {"Orange", 9650.40000}, {"Evanston", 5219.50000} }},
        {"CEM_Commercial_CEM", { {"Watson", 5093.25000}, {"Bayshore", 11355.05000}, {"Orange", 5320.40000}, {"Evanston", 10433.70000}}},
        {"CEM_New_Brunswick", { {"Watson", 4987.72500}, {"Bayshore", 8935.05000}, {"Orange", 7320.40000}, {"Evanston", 5875.50000}}},
        {"CEM_Bernards_PGroup", { {"Watson", 8011.72500}, {"Bayshore", 8300.05000}, {"Orange", 12110.40000}, {"Evanston", 10000.50000}}}
  };

  // Load data into AMPL instances
  load_data(master, FACILITIES, CUSTOMERS, SCENARIOS, facility_capacity_map, customer_demand_map,
    variable_cost_map);
  load_master_data(master);
  load_data(sub, FACILITIES, CUSTOMERS, SCENARIOS, facility_capacity_map, customer_demand_map,
    variable_cost_map);

  // Export to AMPLS
  auto master_ampls = ampls::AMPLAPIInterface::exportModel<T>(master);
  auto map = master_ampls.getVarMap();
  
  master.setOption("solver", master_ampls.driver());
  sub.setOption("solver", master_ampls.driver());

  // Store some maps between the AMPL variables and the position in the solvers view
  std::map<std::string, int> index_facility_open, index_sub_variable_cost;
  std::map<int, std::string> revindex_facility_open, revindex_sub_variable_cost;
  for (auto f : FACILITIES) {
    int in = findIndexWithSubstring(map, f);
    index_facility_open[f] = in;      // map FACILITY -> solverindex facility_open[FACILITY]
    revindex_facility_open[in] = f;   // map solverindex facility_open[FACILITY] -> FACILITY
  }
  for (auto s : SCENARIOS) {
    int in = findIndexWithSubstring(map, s);
    index_sub_variable_cost[s] = in;      // map SCENARIOS -> solverindex sub_variable_cost[SCENARIOS]
    revindex_sub_variable_cost[in] = s;   // map solverindex sub_variable_cost[SCENARIOS] -> CUSTOMER

  }
  
  // Constraints of subproblems that we will use to extract dual/unbounded rays
  auto satisfying_customer_demand = sub.getConstraint("satisfying_customer_demand");
  auto facility_capacity_limits = sub.getConstraint("facility_capacity_limits");
  auto scenario_subproblem = sub.getParameter("sub_scenario");
  
  // Maps to store indices and coefficients (used when actually adding cut using AMPLS)
  int n_noviolations = 0;
  double epsilon = 0.00000001;
  std::map<std::string, double> sub_variable_cost;
  for (auto s : SCENARIOS) {
    sub_variable_cost[s] = 0;
  }
  
  // Set formatting to avoid scientific notation
  std::cout << std::fixed << std::setprecision(2);
  // Main iterations loop
  for (int it = 1; ;it++) {
    std::cout << std::endl << "******* Iteration " << it << " *******";
    n_noviolations = 0;
    for (auto s : SCENARIOS) {
      scenario_subproblem.set(s); // set the scenario in the subproblem
      sub.getOutput("solve;");
      sub.solve();
      auto result = sub.getValue("solve_result").str();
      std::cout << std::endl << std::endl << "Scenario " << s << "   ";
      if (result == "infeasible") {
        auto scd = satisfying_customer_demand.getValues("dunbdd");
        auto fcl = facility_capacity_limits.getValues("dunbdd");
        addBendersCut(master_ampls, scd, fcl,
          index_facility_open, 0, facility_capacity_map,
          customer_demand_map.at(s), false);
      }
      else if (sub.getValue("operating_cost").dbl() > sub_variable_cost[s] + epsilon)
      {
        auto scd = satisfying_customer_demand.getValues("dual");
        auto fcl = facility_capacity_limits.getValues("dual");
        addBendersCut(master_ampls, scd, fcl,
          index_facility_open, index_sub_variable_cost.at(s), facility_capacity_map,
          customer_demand_map.at(s), true);
      }
      else
      {
        std::cout << "No violation." << std::endl;
        n_noviolations += 1;
      }
    }
    // If no scenario violates optimality and feasibility conditions
    // we converged to a solution
    
    if (n_noviolations == SCENARIOS.size()) break;

    // Otherwise resolve the master
    std::cout << "Resolving master problem" << std::endl;
    master_ampls.optimize();
    std::cout << "OBJ=" << master_ampls.getObj() << std::endl;
    // Set sub_facility_open with the master solution
    // Here we need to pass from solver view (indices) to AMPL 
    // view (variable name and index value)
    auto sol = master_ampls.getSolutionVector();
    std::vector<ampl::Tuple> indices;
    std::vector<double> values;
    for (auto a : revindex_facility_open) {
      indices.push_back(ampl::Tuple(a.second));
      values.push_back(sol[a.first]);
    }
    sub.getParameter("sub_facility_open").setValues(indices.data(), values.data(), indices.size());
    // Update the sub_variable_cost values
    for (auto a : revindex_sub_variable_cost)
      sub_variable_cost[a.second] = sol[a.first];
  }
  std::cout << std::endl << std::endl << "***** Optimal solution found, cost: " << master_ampls.getObj() << std::endl;

  // Import the AMPLS model back to AMPL, and show the additional cuts
  ampls::AMPLAPIInterface::importModel(master, master_ampls);
  std::cout << "Optimal solution in AMPL:" << master.getObjective("TotalCost").value() << std::endl;
  master.eval("expand {c in 1.._ncons} _con[c];");

  assert(abs(master.getObjective("TotalCost").value() - 756943875) < 10e-3);
  
}

template <class T> void example() {
  ampl::AMPL master, sub;
  try {
    create_master_problem(master);
    create_sub_problem(sub);
    solve<T>(master, sub);
   
  }
  catch (const std::exception& e) {
    std::cout << e.what();
  }
}

int main(int argc, char** argv) {
  double obj;
#ifdef USE_highs
  example<ampls::HighsModel>();
#endif

#ifdef USE_scip
  example<ampls::SCIPModel>();
#endif

#ifdef USE_gurobi
  example<ampls::GurobiModel>();
#endif




#ifdef USE_xpress
  example<ampls::XPRESSModel>();
#endif

#ifdef USE_cplex
 example<ampls::CPLEXModel>();
#endif

#ifdef USE_copt
  example<ampls::CoptModel>();
#endif

#ifdef USE_cbcmp
 example<ampls::CbcModel>();
#endif

  return 0;
}

File: python/examples/benders.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from calendar import c
from json.encoder import INFINITY
from amplpy import AMPL, variable
import pandas as pd

import amplpy_gurobi as ampls 
SOLVER = "gurobi"

def create_extended_form() ->AMPL:
    a = AMPL()
    a.eval("""
                set FACILITIES; 
                set CUSTOMERS;  
                set SCENARIOS;  

                var facility_open{FACILITIES} binary;                   
                var production{FACILITIES, CUSTOMERS, SCENARIOS} >= 0;  

                param fixed_cost{FACILITIES} >= 0;                
                param variable_cost{FACILITIES, CUSTOMERS} >= 0; 
                param customer_demand{CUSTOMERS, SCENARIOS} >= 0; 
                param facility_capacity{FACILITIES} >= 0;        
                param prob{SCENARIOS} default 1/card(SCENARIOS);  


                minimize TotalCost: 
                    sum{i in FACILITIES} fixed_cost[i] * facility_open[i] +                                                             
                    sum{s in SCENARIOS, i in FACILITIES, j in CUSTOMERS} prob[s] * variable_cost[i,j] * production[i,j,s];  

                s.t. satisfying_customer_demand{s in SCENARIOS, j in CUSTOMERS}:
                    sum{i in FACILITIES} production[i,j,s] >= customer_demand[j,s];

                s.t. facility_capacity_limits{s in SCENARIOS, i in FACILITIES}:
                    sum{j in CUSTOMERS} production[i,j,s] <= facility_capacity[i] * facility_open[i];

                s.t. sufficient_production_capacity:
                    sum{i in FACILITIES} facility_capacity[i]*facility_open[i] >= max{s in SCENARIOS} sum{j in CUSTOMERS} customer_demand[j,s];  
           """)
    a.option["solver"]=SOLVER
    return a

def set_data(a: AMPL, is_master: bool):
    a.eval("""data; set FACILITIES  := Watson Bayshore Orange Evanston;
            set CUSTOMERS   := CEM_Store_41594 CEM_Cypress_Market CEM_Time_Mart_23 CEM_Cypress_CEM CEM_Fuel_Maxx_84 CEM_Commercial_CEM CEM_New_Brunswick CEM_Bernards_PGroup;
            set SCENARIOS   := Low Medium High;
            param variable_cost (tr):       Watson      Bayshore        Orange              Evanston    :=
                    CEM_Store_41594         6739.72500  10355.05000     7650.40000          5219.50000
                    CEM_Cypress_Market      4739.72500  8355.05000      17320.40000         15000.50000
                    CEM_Time_Mart_23        7739.72500  9355.05000      7320.40000          5000.50000
                    CEM_Cypress_CEM         5739.72500  6355.05000      7320.40000          9433.70000
                    CEM_Fuel_Maxx_84        9739.72500  10355.05000     9650.40000          5219.50000
                    CEM_Commercial_CEM      5093.25000  11355.05000     5320.40000          10433.70000
                    CEM_New_Brunswick       4987.72500  8935.05000      7320.40000          5875.50000
                    CEM_Bernards_PGroup     8011.72500  8300.05000      12110.40000         10000.50000;

            param customer_demand:          Low         Medium          High    :=
                    CEM_Store_41594         10000       15000           20000
                    CEM_Cypress_Market      10000       15000           20000
                    CEM_Time_Mart_23        10000       15000           20000
                    CEM_Cypress_CEM         10000       15000           20000
                    CEM_Fuel_Maxx_84        10000       15000           20000
                    CEM_Commercial_CEM      10000       15000           20000
                    CEM_New_Brunswick       10000       15000           20000
                    CEM_Bernards_PGroup     10000       15000           20000;

            param facility_capacity := Watson 75000 Bayshore 200000 Orange 150000 Evanston 90000;
            """)
    if is_master:
        a.eval("param fixed_cost := Watson 39268000 Bayshore 65268000 Orange 60268000 Evanston 48268000;")
   
def create_common() -> AMPL:
    a = AMPL();
    a.option["solver"]="cplex"
    a.option["cplex_options"]="presolve 0 dualopt"
    a.eval("set FACILITIES;"
    "set CUSTOMERS;"
    "set SCENARIOS;"
    "param prob{SCENARIOS} default 1/card(SCENARIOS);"
    "param sub_scenario symbolic in SCENARIOS;"
    "param customer_demand{CUSTOMERS, SCENARIOS} >= 0;"

    "param facility_capacity{FACILITIES} >= 0;"
    "param variable_cost{ FACILITIES, CUSTOMERS } >= 0;")
    return a
def create_sub_problem() ->AMPL:
  a = create_common()
  a.eval("var production{FACILITIES, CUSTOMERS, SCENARIOS} >= 0;"
        "param sub_facility_open{ FACILITIES } default 1; "
        "minimize operating_cost:"
        "sum{ i in FACILITIES, j in CUSTOMERS }"
        "variable_cost[i, j] * production[i, j, sub_scenario]; "

        "s.t.satisfying_customer_demand{ j in CUSTOMERS }:"
        "sum{i in FACILITIES} production[i, j, sub_scenario] >= customer_demand[j, sub_scenario];"

        "s.t.facility_capacity_limits{i in FACILITIES}:"
        "sum{ j in CUSTOMERS } production[i, j, sub_scenario] <= facility_capacity[i] * sub_facility_open[i];")
  return a

def create_master_problem() ->AMPL:
  a=create_common()
  a.eval(
      """param fixed_cost{ FACILITIES } >= 0;
         var sub_variable_cost{SCENARIOS} >= 0;
         var facility_open{ FACILITIES } binary;
         minimize TotalCost :
            sum{ i in FACILITIES } fixed_cost[i] * facility_open[i] + sum{s in SCENARIOS} prob[s]*sub_variable_cost[s];
         s.t. sufficient_production_capacity:
            sum{i in FACILITIES} facility_capacity[i]*facility_open[i] >= max{s in SCENARIOS} sum{j in CUSTOMERS} customer_demand[j,s];
        """)
  return a


def get_data_from_ampl(a: AMPL) -> tuple:
    """Get data we need in the algorithm from AMPL"""
    facilities=a.set["FACILITIES"].get_values().to_list()
    customers=a.set["CUSTOMERS"].get_values().to_list()
    scenarios=a.set["SCENARIOS"].get_values().to_list()
    facility_capacity=a.param["facility_capacity"].get_values().to_dict()
    customer_demand = a.param["customer_demand"].get_values().to_dict()
    variable_cost = a.param["variable_cost"].get_values().to_dict()
    return (facilities, customers, scenarios, facility_capacity, customer_demand, variable_cost)
    
def get_tuple_map(model: ampls.AMPLModel, varname: str):
    '''Get a map indexingsetitem -> index for a specified variable'''
    var_map = model.getVarMapFiltered(varname)
    beg = len(varname) + 2
    return { name[beg:-2] : value for name,value in var_map.items()} 

def get_maps(model) -> tuple:
    '''Get all the maps we will need for the execution'''
    index_facility_open = get_tuple_map(model, "facility_open")
    index_sub_variable_cost = get_tuple_map(model, "sub_variable_cost")
    revindex_facility_open = {value : name for name,value in index_facility_open.items()}
    revindex_sub_variable_cost = {value : name for name,value in index_sub_variable_cost.items()}
    return (index_facility_open, revindex_facility_open, index_sub_variable_cost, revindex_sub_variable_cost)

   
def add_benders_cut(model: ampls.AMPLModel, cdduals: dict, fcduals: dict, 
                    index_facility_open: dict, index_sub_variable_cost: dict,
                    facility_capacity: dict, customer_demand: dict, 
                    s: str, customers: list, cutype: str):
    """Add a Bender's cut to the model."""
    # Note that to accomodate for solver view, we formulate it as below
    # sub_variable_cost[s]-sum{i in FACILITIES} facility_price[i,s,k]*facility_capacity[i]*facility_open[i]
    # >=   sum{j in CUSTOMERS} customer_price[j,s,k]*customer_demand[j,s]; 
    indices = [index_facility_open[f] for f in fcduals.keys()]
    coeffs = [-fcduals[f] * facility_capacity[f] for f in fcduals.keys()]
    if cutype=="optimality":
        indices.append(index_sub_variable_cost[s])
        coeffs.append(1)
    rhs = sum(cdduals[c] * customer_demand[(c,s)] for c in customers)
    opt = model.addConstraint(indices, coeffs, ampls.CutDirection.GE, rhs)
    # With the following, we keep track of the cuts we add, to then add them back
    # into AMPL
    model.record(opt);
    print(f"Added {cutype} cut: {opt.toString()}\n")
    
def doStuff(set_data_function):
    '''Generic function doing the optimization and reading the results'''
    master=create_master_problem()
    sub=create_sub_problem()
    
    set_data_function(sub, False)
    set_data_function(master, True)
    
    # Get the data we will use for iterations
    (facilities, customers, scenarios, facility_capacity, customer_demand, variable_cost)=get_data_from_ampl(master)
    
    # Set some options
    master.option["presolve"]=0
    sub.eval("suffix dunbdd;")
    sub.option["presolve"]=0 # set else AMPL likely presolves the variable facility_open
    
    # Export the master problem to ampls
    master_ampls = master.to_ampls(SOLVER)
    # Get maps between solver vars and AMPL entities
    (index_facility_open, revindex_facility_open, index_sub_variable_cost, revindex_sub_variable_cost)=get_maps(master_ampls)
    
    # Constraints of subproblems that we will use to extract dual/unbounded rays
    # Could do without but this makes the implementation slightly cleaner
    satisfying_customer_demand = sub.con["satisfying_customer_demand"]
    facility_capacity_limits = sub.con["facility_capacity_limits"]
    
    # Initialize Benders's counter and params
    epsilon = 0.00000001
    sub_variable_cost = {s : 0 for s in scenarios}
    it =0
    while True:
       it+=1
       print(f"****** Iteration {it} *******")
       nviolations = 0
       for s in scenarios:
           # Solve the subproblem
           sub.param["sub_scenario"]=s # Assign scenario
           sub.get_output("solve;")
           result = sub.get_value("solve_result")
           print(f"Scenario {s} Objective={sub.get_current_objective().value()}")
           # Decide what cut to add, if any
           if (result == "infeasible"):
                cdduals = satisfying_customer_demand.get_values("dunbdd").to_dict();
                fcduals = facility_capacity_limits.get_values("dunbdd").to_dict()
                add_benders_cut(master_ampls, cdduals, fcduals, index_facility_open, index_sub_variable_cost,
                                    facility_capacity, customer_demand, s, customers, "feasibility")
           elif sub.getValue("operating_cost") > sub_variable_cost[s] + epsilon:
               cdduals = satisfying_customer_demand.get_values("dual").to_dict()
               fcduals = facility_capacity_limits.get_values("dual").to_dict()
               add_benders_cut(master_ampls, cdduals, fcduals, index_facility_open, index_sub_variable_cost,
                                    facility_capacity, customer_demand, s, customers, "optimality")
           else:
               nviolations+=1
               
       # It no scenario had a violation, we exit
       if nviolations == len(scenarios):
           break;
       
       print("SOLVING MASTER PROBLEM")
       master_ampls.optimize();
    
       # Get the solution vector from ampls
       sol = master_ampls.getSolutionVector()
       # AMPL: let {f in FACILITIES} sub_facility_open[f] := facility_open[f];
       sub.param["sub_facility_open"]={f : sol[i] for i,f in revindex_facility_open.items()}
       # Assign the costs from the master problem for next iteration
       sub_variable_cost =            {s : sol[i] for i,s in revindex_sub_variable_cost.items()}
    # End of Bender's loop
   
    print(f"Optimal solution found, cost: {master_ampls.get_obj()}\n\n")
    # Import the AMPLS model back to AMPL, and show the additional cuts
    master.import_solution(master_ampls, keep_files=True)
    master.eval(master_ampls.getRecordedEntities())
    master_obj=master.get_current_objective().value()
    master.eval("expand {c in 1.._ncons} _con[c];")
    print(f"Optimal solution benders: {master_obj}")    
    
    # Check with the extended form
    extended = create_extended_form()
    set_data_function(extended, True)
    extended.solve()
    extended_obj = extended.get_current_objective().value()
    print(f"Optimal solution extended: {extended_obj}")
    if abs(extended_obj-master_obj)>10e-4:
        print(f"Not good at all, discrepancy of {abs(extended_obj-master_obj)} found!!!")
        print(f"extended_obj={extended_obj} master_obj={master_obj}")
        
doStuff(set_data)